File: Graph Theory Peaks/Graph/Q2. Add Edges to Make Degrees of All Nodes Even/Q2_AddEdgestoMakeDegreesofAllNodesEven.py
class Solution(object):
    def isPossible(self, n, edges):
        
        # Create adjacency list (graph) using sets
        G = [set() for i in range(n)]
        
        # Build undirected graph (convert to 0-based index)
        for i, j in edges:
            G[i-1].add(j-1)  # add edge i -> j
            G[j-1].add(i-1)  # add edge j -> i
        
        # Find nodes with odd degree
        odd = [i for i in range(n) if len(G[i]) % 2]

        # Helper function:
        # Returns True if there is NO edge between a and b
        def f(a, b):
            return a not in G[b]

        # Case 1: If exactly 2 nodes have odd degree
        if len(odd) == 2:
            a, b = odd
            
            # Try connecting both odd nodes to some third node
            # such that new edges do not already exist
            return any(f(a, i) and f(b, i) for i in range(n))

        # Case 2: If exactly 4 nodes have odd degree
        if len(odd) == 4:
            a, b, c, d = odd
            
            # Try all possible pairings of 4 nodes
            return  f(a, b) and f(c, d) or \
                    f(a, c) and f(b, d) or \
                    f(a, d) and f(c, b)

        # Case 3: If no odd nodes → already valid
        return len(odd) == 0

File: Graph Theory Peaks/Graph/Q2. Add Edges to Make Degrees of All Nodes Even/Q2_AddEdgestoMakeDegreesofAllNodesEven.java
class Solution {
    public boolean isPossible(int n, List<List<Integer>> edges) {
        
        // Adjacency list using HashSet (for fast lookup)
        List<Set<Integer>> l = new ArrayList<>();
        
        // Array to store degree of each node (1-based indexing)
        int degree[] = new int[n+1];
        
        // Initialize adjacency list
        for(int i = 0; i < n+1; i++){
            l.add(new HashSet<>());
        }

        // Build graph and count degrees
        for(List<Integer> x : edges){
            l.get(x.get(0)).add(x.get(1));  // add edge u -> v
            l.get(x.get(1)).add(x.get(0));  // add edge v -> u
            degree[x.get(0)]++;             // increase degree of u
            degree[x.get(1)]++;             // increase degree of v
        }

        // Store nodes having odd degree
        List<Integer> odd = new ArrayList<>();
        for(int i = 1; i < n+1; i++){
            if(degree[i] % 2 == 1){  // if degree is odd
                odd.add(i);
            }
        }

        // If all nodes have even degree → already valid
        if(odd.size() == 0) return true;

        // If exactly 2 nodes have odd degree
        if(odd.size() == 2){
            
            Set<Integer> s1 = l.get(odd.get(0));
            Set<Integer> s2 = l.get(odd.get(1));

            // If they are not directly connected → add edge between them
            if(!s1.contains(odd.get(1))){
                return true;
            }
            // If already connected → try connecting both to another node
            else{
                for(int i = 1; i < n+1; i++){
                    if(!s1.contains(i) && !s2.contains(i)){
                        return true;
                    }
                }
            }
        }

        // If exactly 4 nodes have odd degree
        if(odd.size() == 4){
            
            Set<Integer> s1 = l.get(odd.get(0));
            Set<Integer> s2 = l.get(odd.get(1));
            Set<Integer> s3 = l.get(odd.get(2));
            Set<Integer> s4 = l.get(odd.get(3));

            // Case 1: pair (1,2) and (3,4)
            if(!s1.contains(odd.get(1)) && !s3.contains(odd.get(3))){
                return true;
            }
            // Case 2: pair (1,3) and (2,4)
            else if(!s1.contains(odd.get(2)) && !s2.contains(odd.get(3))){
                return true;
            }
            // Case 3: pair (1,4) and (2,3)
            else if(!s1.contains(odd.get(3)) && !s2.contains(odd.get(2))){
                return true;
            }
        }

        // If none of the above conditions satisfied → not possible
        return false;
    }
}